sdks/java/io/elasticsearch-tests/elasticsearch-tests-common/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTestCommon.java

```diff
@@ -460,6 +462,40 @@ void testWriteWithErrorsReturnedAllowedErrors() throws Exception {
     pipeline.run();
   }
 
+  void testWriteWithElasticClientResponseException() throws Exception {
+    Write write =
+        ElasticsearchIO.write()
+            .withConnectionConfiguration(connectionConfiguration)
+            .withMaxBatchSize(1000L)
+            .withMaxBatchSizeBytes(
+                Long.MAX_VALUE) // Max long number to make sure it flushes 2 docs.
+            .withThrowWriteErrors(false)
+            .withIdFn(new ExtractValueFn("id"))
+            .withUseStatefulBatches(true);
+
+    List<String> data =
+        ElasticsearchIOTestUtils.createDocuments(
+            numDocs, ElasticsearchIOTestUtils.InjectionMode.INJECT_ONE_ID_TOO_LONG_DOC);
+
+    PCollectionTuple outputs = pipeline.apply(Create.of(data)).apply(write);
+    PCollection<String> success =
+        outputs
+            .get(Write.SUCCESSFUL_WRITES)
+            .apply("Convert success to input ID", MapElements.via(mapToInputIdString));
+
+    PCollection<String> fail =
+        outputs
+            .get(Write.FAILED_WRITES)
+            .apply("Convert fails to input ID", MapElements.via(mapToInputIdString));
+
+    Set<String> failedIds =
+        IntStream.range(0, data.size() - 1).mapToObj(String::valueOf).collect(Collectors.toSet());
+    failedIds.add(INVALID_LONG_ID);
+    PAssert.that(success).empty();
+    PAssert.that(fail).containsInAnyOrder(failedIds);
```

Comment:
Looking awesome, I think this is the last note:

Could you also add an assert to verify the contents of the `doc` elements in `fail` ?  Ex. that `hasError` is truthy, and that `responseItem` is useful/readable?

Reply:
actually I suppose just verifying the contents of the response item, because hasError has to be truthy to be in the `fail` PCollection

Reply:
```
String errorJsonMessage =
              String.format(
                  "{\"message\":\"java.io.IOException was thrown in batch flush: %s\"}",
                  elasticResponseExceptionMessage);
```
Revised the error message from containing `org.elasticsearch.client.ResponseException` to `java.io.IOException` since we are now handling things within java.io.IOException.

I am asserting the error message contains `java.io.IOException was thrown in batch flush`. I think this is good enough. The rest of the message comes directly from the exception message.